Update operation docs (#1802)
graphql/graphql.github.io: src/pages/learn/mutations.mdx

import { Callout } from "nextra/components"

# Mutations

<p className="learn-subtitle">Learn how to modify data with a GraphQL server</p>

Most discussions of GraphQL focus on data fetching, but any complete data platform needs a way to modify server-side data as well.

> In REST, any request might cause some side-effects on the server, but by convention, it's suggested that one doesn't use `GET` requests to modify data. GraphQL is similar—technically any query could be implemented to cause a data write. However, it's useful to establish a convention that any operations that cause writes should be sent explicitly via a mutation.

Reviewer: ```suggestion
In REST, any request might cause some side-effects on the server, but by convention, it's suggested that one doesn't use `GET` requests to modify data. GraphQL is similar—technically any field resolver could be implemented to cause a data write—but the [GraphQL specification states](https://spec.graphql.org/draft/#sel-GANRNDAB6DBmMn6D):

> the resolution of fields other than top-level mutation fields must always be side effect-free and idempotent

Thus only the top-level fields in mutation operations are allowed to cause side effects; to do otherwise would make your schema GraphQL non-compliant.
```

> On this page, you'll learn how to use mutation operations to create, update, and delete data using GraphQL.

Reviewer: I am very uncomfortable with this positioning; GraphQL isn't just a CRUD system and in general it's better to use semantically named mutations for specific user actions rather than trying to fit every mutation into the CRUD box. Some mutations such as `processPayment`, `resetPassword`, `sendEmailVerification`, `inviteUser`, `markAllNotificationsAsRead`, `redeemGiftCard` and so on (arguably) don't even fit into the CRUD box. (This isn't to say that CRUD should never be used, but more that a semantically named action should be exposed preferentially where it makes sense to do so.)

From GraphQL's point of view there are no different types of mutation - there is no difference between a create, update, delete or arbitrary user action mutation - so I'm concerned that highlighting CRUD mutations in this way sets the reader (especially someone who does not read the page fully) up for misconceptions.

Author: I see where you're coming from. I think the challenge here is to figure out how much information to include up-front in the intro for a reader who has some experience with REST and needs to connect those dots with GraphQL (e.g. #992). This distinction is called out in the paragraph that follows the new example for updating a human name, but it's halfway down the page.

Do you think it would be better to generalize it like this?

> On this page, you'll learn how to use mutation operations to modify data on a server using GraphQL.

Author: Do you think the subheads on the page are problematic? We could further deemphasize any kind of 1:1 CRUD implication by renaming them like this:

~Create data~ Add new data
~Update data~ Update existing data
~Delete data~ Remove existing data

Author: One more thought on this is—I think there's a gap in the Best Practices content where "migrating from REST" could be covered in more detail too with a few more detailed examples that are based on a further-enhanced version of the Star Wars schema.

Reviewer: I'm not sure that our learn docs should necessarily cater primarily to people who already know REST/CRUD - perhaps a separate guide for "how to convert CRUD REST operations to GraphQL" might make sense but I think our learn docs shouldn't start with REST concepts and map them to GraphQL, but instead start with GraphQL best practices and then if necessary expand with REST accommodations/analogies.

I think my preferred way of addressing this would be to start with a non-CRUD mutation ~~that follows best practices (input object, mutation payload)~~ so that that is front and centre, and then maybe have a sub heading about CRUD that has examples of create, update, upsert, delete. I think the original documentation used a create mutation because that's fairly universal, but perhaps pivoting away from that would help learners to do GraphQL the right way from the start?

Trying to think of something within the Star Wars schema that would make sense as a non-CRUD mutation, perhaps `rateFilm` which stores a rating "onto the film" from your perspective, but clearly isn't actually updating the film itself:

```graphql
type Mutation {
  rateFilm(input: RateFilmInput!): RateFilmPayload
}

input RateFilmInput {
  filmId: ID!
  rating: FilmRating! = THUMBS_UP
}

type RateFilmPayload {
  film: Film
}

type Film {
  # ...
  viewerRating: FilmRating
}

enum FilmRating {
  THUMBS_DOWN
  NEUTRAL
  THUMBS_UP
  DOUBLE_THUMBS_UP
}
```

Reviewer: Actually I think this page should talk about the input object/payload best practices at the end, so maybe simplify this down:

```graphql
type Mutation {
  rateFilm(
    filmId: ID!
    rating: FilmRating! = THUMBS_UP
  ): Film
}

type Film {
  # ...
  viewerRating: FilmRating
}

enum FilmRating {
  THUMBS_DOWN
  NEUTRAL
  THUMBS_UP
  DOUBLE_THUMBS_UP
}
```

Author: I can come up with something but at this point I think I'll wait until #1812 is merged so I rebase the changes made in `swapi-schema.tsx` in that PR first.

Reviewer: - #1812

is now merged :+1:

Author: There's a new commit with the requested changes included now.

mutation UpdateHumanName($id: ID!, $name: String!) { 
  updateHumanName(id: $id, name: $name ) {
    id
    name
  }
}
```

This example demonstrates an important distinction from REST. To update a human's properties using a REST API, you would likely send any updated data to a generalized endpoint for that resource using a `PATCH` request. With GraphQL, instead of simply creating an `updateHuman` mutation, you can define more specific mutation fields such as `updateHumanName`that are designed for the task at hand.

Purpose-built mutation fields can help make a schema more expressive by allowing the input types for field arguments to be Non-Null types (a generic `updateHuman` mutation would likely need to accept many nullable arguments to handle different update scenarios). Defining this requirement in the schema also eliminates the need for other runtime logic to determine that the appropriate values were submitted to perform the client's desired write operation.

Schemas should be designed to help clients get the data that they need from the GraphQL API, so the fields defined in a schema should be informed by those use cases.

## Deleting data

> Just as we can send a `DELETE` request to delete a resource with a REST API, we can use mutations to delete some existing data as well by defining another field on the `Mutation` type:

Reviewer 2: I think we should clarify that in GraphQL we send a `POST` request to delete

Author: I have an updated "Serving over HTTP" page almost ready to go where I go into this in detail based on the draft GraphQL over HTTP spec (including when `GET` may be used too). Does it make sense to duplicate that here? My goal was draw connections with REST patterns the reader would likely be familiar with while explaining how different kinds of mutations work without overloading with too much information. It can be tricky to find the right balance though! Let me know what you think.

On this page, you'll learn how to use mutation operations to write data using GraphQL, and do so in a way that supports client use cases.

<Callout type="info">
All of the features of GraphQL operations that apply to queries also apply to mutations, so review the [Queries](/learn/queries/) page first before proceeding.
</Callout>

## Add new data

When creating new data with a REST API, you would send a `POST` request to a specific endpoint and include information about the entities to be created in the body of the request. GraphQL takes a different approach.

Let's look at an example mutation that's defined in our schema:

```graphql
enum Episode {
  NEWHOPE
  EMPIRE
  JEDI
}

input ReviewInput {
  stars: Int!
  commentary: String
}

type Mutation {
  createReview(episode: Episode, review: ReviewInput!): Review
}
```

Like queries, mutation fields are added to one of the [root operation types](https://spec.graphql.org/draft/#sec-Root-Operation-Types) that provide an entry point to the API. In this case, we define the `createReview` field on the `Mutation` type. 

> Mutation fields can also accept arguments and you might notice that the `review` argument has an input type set to `ReviewInput`. This is known as [Input Object type](/learn/schema/#input-object-types), which allows us to pass in a whole object to be created instead of individual scalar values only.

Reviewer: Let's not imply that the input object itself is "created":

```suggestion
Mutation fields can also accept arguments and you might notice that the `review` argument has an input type set to `ReviewInput`. This is known as [Input Object type](/learn/schema/#input-object-types), which allows us to pass in a structured object containing information the mutation can use instead of individual scalar values only.
```

Also like just like queries, if the mutation field returns an Object type, then you specify a selection set of its fields in the operation:

```graphql
# { "graphiql": true, "variables": { "ep": "JEDI", "review": { "stars": 5, "commentary": "This is a great movie!" } } }
mutation CreateReviewForEpisode($ep: Episode!, $review: ReviewInput!) {
  createReview(episode: $ep, review: $review) {
    stars
    commentary
  }
}
```

> While the `createReview` field could be defined with any valid output type in the schema, it's conventional to specify an output type that corresponds to whatever is modified during the mutation—in this case, the `Review` type. This can be useful for clients that need to fetch the new state of an object after an update.

Reviewer: I don't want to imply that `createReview(...): Review` is the convention people should follow when `createReview(...): CreateReviewPayload` is a much better pattern. I've changed "corresponds" to "relates" since the `CreateReviewPayload` and `Review` both _relate_ to the type that's modified, whereas I'm not sure they both "correspond" to it? (I'm no English scholar though!)

```suggestion
While the `createReview` field could be defined with any valid output type in the schema, it's conventional to specify an output type that relates to whatever is modified during the mutation—in this case, the `Review` type. This can be useful for clients that need to fetch the new state of an object after an update.
```

Recall that GraphQL is meant to work with your existing code and data, so the actual creation of the review is up to you when clients send this operation to the GraphQL server. A hypothetical function that writes the new review to a database during a `createReview` mutation might look like this:

```js
Mutation: {
  createReview(obj, args, context, info) {
    return context.db
      .createNewReview(args.episode, args.review)
      .then((reviewData) => new Review(reviewData))
  }
}
```

> You can learn more about GraphQL provides data for fields on the [Execution page](/learn/execution).

Reviewer: Missing word?

```suggestion
You can learn more about how GraphQL provides data for fields on the [Execution page](/learn/execution).
```

## Update existing data

Similarly, we use mutations to update existing data. To change a human's name, we'll define a new mutation field and set that field's output type to the `Human` type so we can return the updated human's information to client after the server successfully writes the data:

```graphql
type Mutation {
  updateHumanName(id: ID!, name: String!): Human
}
```

This operation will update Luke Skywalker's name:

```graphql
# { "graphiql": true, "variables": { "id": "1000", "name": "Luke Starkiller" } }
mutation UpdateHumanName($id: ID!, $name: String!) { 
  updateHumanName(id: $id, name: $name ) {
    id
    name
  }
}
```

> This example demonstrates an important distinction from REST. To update a human's properties using a REST API, you would likely send any updated data to a generalized endpoint for that resource using a `PATCH` request. With GraphQL, instead of simply creating an `updateHuman` mutation, you can define more specific mutation fields such as `updateHumanName`that are designed for the task at hand.

Reviewer: This is already really good, but I've edited it to make it even clearer this is a choice.

```suggestion
This example demonstrates an important distinction from REST. To update a human's properties using a REST API, you would likely send any updated data to a generalized endpoint for that resource using a `PATCH` request. With GraphQL, instead of creating a generic `updateHuman` mutation, you may choose to define more specific mutation fields such as `updateHumanName` that are designed for the task at hand.
```

Reviewer: Should this section have it's own header: `## Purpose-driven mutations`

Purpose-built mutation fields can help make a schema more expressive by allowing the input types for field arguments to be Non-Null types (a generic `updateHuman` mutation would likely need to accept many nullable arguments to handle different update scenarios). Defining this requirement in the schema also eliminates the need for other runtime logic to determine that the appropriate values were submitted to perform the client's desired write operation.

GraphQL also allows us to express relationships between data that would be more difficult to model semantically with a basic CRUD-style request. For example, a user may wish to save a personal rating for a film. While the rating belongs to the user and doesn't modify anything related to a film itself, we can ergonomically associate it with a `Film` object as follows:

```graphql
# { "graphiql": true, "variables": { "episode": "EMPIRE", "rating": "THUMBS_UP" } }
mutation RateFilm($episode: Episode!, $rating: FilmRating!) { 
  rateFilm(episode: $episode, rating: $rating) {
    episode
    viewerRating
  }
}
```

> As a general rule, schemas should be designed to help clients get the data that they need from the GraphQL API, so the fields defined in a schema should be informed by those use cases.

Reviewer: This is really great content, but "get the data that they need from" seems very query-focussed (and side-effect free) - can you tweak it so that it better reflects mutations?

Author: Fixed in https://github.com/graphql/graphql.github.io/pull/1802/commits/1f3ae06e30c231b965c8f591ababce6d9c88ae8f.

Reviewer: Perfect 🙌
